### PR TITLE
Add OMC-1016 [Nimbus] "Open Share Sheet" Card Messaging Action

### PR DESCRIPTION
**Pull request by halemu**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/OMC-1016)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/22967)

## :bulb: Description
Adds a new messaging action to open the share sheet from a message. The message is configurable using Nimbus and contains a url and title field.

## :pencil: Checklist
You have to check all boxes before merging
- [x] Filled in the above information (tickets numbers and description of your work)
- [x] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [x] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [x] If needed, I updated documentation / comments for complex code and public methods
- [ ] If needed, added a backport comment (example `@Mergifyio backport release/v120`)

### CODE REVIEW

**Comment by halemu** [comment]
This is a work in progress, looking for feedback to ensure I'm heading in the right direction! 🙏 
I'm not able to test it fully since that requires a developer account setup, I'd appreciate testing this using an iPhone as well, thanks!

**Comment by adudenamedruby** [comment]
@halemu I changed this to draft until it's ready, just so it's less confusing for review priorities. :)
Feel free to mark as `Ready for Review` whenever you're done.

**Comment by adudenamedruby** [comment]
As a first quick pass, this is gonna fail on swfitlint due to indentation, so you can either do a swiftlint pass on it, or just do a `re-indent` in the files that you changed.

File: firefox-ios/Client/Coordinators/Router/Route.swift [hunk 1]
```diff
@@ -65,6 +65,7 @@ enum Route: Equatable {
     /// - Parameter section: An instance of `DefaultBrowserSection` indicating the section of the default browser
     ///                      settings to be displayed.
     case defaultBrowser(section: DefaultBrowserSection)
+    case sharesheet(url: URL, title: String?)
```
Comment (adudenamedruby) [inline firefox-ios/Client/Coordinators/Router/Route.swift:68]: we should add documentation for this, like the rest of the functions defined here.

**Comment by halemu** [comment]
Thanks, updated 👍

File: firefox-ios/Client/Coordinators/Router/RouteBuilder.swift [hunk 2]
```diff
@@ -121,6 +121,17 @@ final class RouteBuilder {
 
             case .fxaSignIn:
                 return nil
+
+            case .sharesheet:
+                let linkString = urlScanner.value(query: "url")
+                let titleText = urlScanner.value(query: "title")
+                // Convert the linkString (String?) to a URL? using URL(string:)
```
Comment (adudenamedruby) [inline firefox-ios/Client/Coordinators/Router/RouteBuilder.swift:128]: nit: not really needed, as the code is pretty self documenting here

**Review by adudenamedruby — CHANGES_REQUESTED** [review]

**Review by adudenamedruby — APPROVED** [review]

**Comment by halemu** [comment]
Would appreciate if someone can merge this in since I don't have write access yet, thanks!

**Comment by halemu** [comment]
@Mergifyio backport release/v133

Requesting backport to support iOS Sharesheet experiment.

**Comment by adudenamedruby** [comment]
I've reached out for details.

**Comment by adudenamedruby** [comment]
@Mergifyio backport release/v133

**Pull request merged**